### ISSUE TITLE
Inconsistent computation of gradient in MaxUnPooling

### ISSUE THREAD

**Issue by mergian** [pytorch/pytorch#80827]
### 🐛 Describe the bug

Hey

I think there is a inconsistent definition how MaxUnPool and it's gradient are computed. Here is an example code.

```python
import torch

torch.manual_seed(123)

A = torch.rand(1, 1, 9, 9)

_, I = torch.nn.MaxPool2d(3, 1, return_indices=True)(A)
print("Indices", I)

B = torch.arange(I.numel(), 0, -1).to(torch.float).view(I.shape).detach()
B.requires_grad = True
print("MaxUnPool Input", B)

C = torch.nn.MaxUnpool2d(3, 1)(B, I)
print("MaxUnPool Output", C)

D = C * torch.arange(C.numel()).to(torch.float).view(C.shape)

# now compute the gradient
E = D.sum()
E.backward()

print("MaxUnPool Gradient", B.grad)
```

The output is:
```
Indices tensor([[[[20, 20, 20, 23, 23, 23, 17],
          [28, 30, 30, 23, 23, 23, 34],
          [28, 30, 30, 23, 23, 23, 43],
          [28, 48, 48, 48, 49, 43, 43],
          [47, 48, 48, 48, 49, 43, 43],
          [47, 48, 48, 48, 49, 70, 70],
          [74, 57, 57, 57, 69, 70, 70]]]])
MaxUnPool Input tensor([[[[49., 48., 47., 46., 45., 44., 43.],
          [42., 41., 40., 39., 38., 37., 36.],
          [35., 34., 33., 32., 31., 30., 29.],
          [28., 27., 26., 25., 24., 23., 22.],
          [21., 20., 19., 18., 17., 16., 15.],
          [14., 13., 12., 11., 10.,  9.,  8.],
          [ 7.,  6.,  5.,  4.,  3.,  2.,  1.]]]], requires_grad=True)
MaxUnPool Output tensor([[[[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., 43.],
          [ 0.,  0., 47.,  0.,  0., 37.,  0.,  0.,  0.],
          [ 0., 35.,  0., 33.,  0.,  0.,  0., 36.,  0.],
          [ 0.,  0.,  0.,  0.,  0.,  0.,  0., 15.,  0.],
          [ 0.,  0., 14., 18., 17.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  5.,  0.,  0.,  0.,  0.,  0.],
          [ 0.,  0.,  0.,  0.,  0.,  0.,  3.,  8.,  0.],
          [ 0.,  0.,  7.,  0.,  0.,  0.,  0.,  0.,  0.]]]],
       grad_fn=<MaxUnpool2DBackward0>)
MaxUnPool Gradient tensor([[[[20., 20., 20., 23., 23., 23., 17.],
          [28., 30., 30., 23., 23., 23., 34.],
          [28., 30., 30., 23., 23., 23., 43.],
          [28., 48., 48., 48., 49., 43., 43.],
          [47., 48., 48., 48., 49., 43., 43.],
          [47., 48., 48., 48., 49., 70., 70.],
          [74., 57., 57., 57., 69., 70., 70.]]]])
```

In "Indices" we can see that the index 20 is used 3 times.

Now in "MaxUnPool Output" we see that the index 20 does only use the very last value 47 in this case. So I would expect the gradient for these three to be [0, 0, 20].

I put in a multiplication, so that in the gradient we get the actual index where the value was taken from. As we can see, the gradient for these three values is instead [20, 20, 20].

To wrap this up. To my understanding MaxUnPool computes the following during the forward pass:

```c++
for(int batch ...)
	for(int channel ...)
		for(int index ...)
			output[batch, channel, indices[index]] = input[batch, channel, index];
```

However, when we look at the gradient, all get the value propagated, which is inconsistent with the forward pass, where only the last gets used.

This seems to be inconsistent to me. So I think either the forward pass needs to use a ```+=``` instead of ```=```, or the backward pass needs to only propagate the value to the last occurrence of the index.

Best

### Versions

Collecting environment information...
PyTorch version: 1.12.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 10.3.0
Clang version: Could not collect
CMake version: version 3.23.1
Libc version: glibc-2.17

Python version: 3.7.12 (default, Feb  6 2022, 20:29:18)  [GCC 10.2.1 20210130 (Red Hat 10.2.1-11)] (64-bit runtime)
Python platform: Linux-3.10.0-1160.59.1.el7.x86_64-x86_64-with-centos-7.9.2009-Core
Is CUDA available: False
CUDA runtime version: 11.3.109
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] torch==1.12.0
[pip3] torchmetrics==0.9.1
[pip3] torchvision==0.13.0
[conda] Could not collect

**Comment by Lezcano** [comment]
Could you check whether this function passes [`gradcheck`](https://pytorch.org/docs/stable/generated/torch.autograd.gradcheck.html) with your inputs of choice?

**Comment by nikitaved** [comment]
gradcheck will fail with repeated indices, I am afraid, that is why OpInfo tests make sure that there are no repeated indices. I am afraid the kernel does not take into consideration any repeated indices at all.

**Comment by mergian** [comment]
I agree with @nikitaved, the documentation of gradcheck explicitly says that it does not work with repeated indices.

Here is code taken from ```ATen/native/cpu/MaxUnpoolKernel.cpp``` (I removed unnecessary lines for simplicity)

### Forward:
```cpp
at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
	for (const auto i : c10::irange(begin, end)) {
		scalar_t* output_ptr = output_data + c * output_image_size;

		int64_t maxp = indices_data[i];
		output_ptr[maxp] = input_data[i];

		// move on to next input index
		data_index_step(c, channels, ip, input_image_size);
	}
});
```

### Backward:
```cpp
at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
	for (const auto i : c10::irange(begin, end)) {
		scalar_t* grad_output_ptr = grad_output_data + c * output_image_size;

		int64_t maxp = indices_data[i];
		grad_input_data[i] = grad_output_ptr[maxp];

		// move on to next input index
		data_index_step(c, channels, ip, input_image_size);
	}
});
```

So the forward pass overrides all previous values that use the same index. This looks to be pretty random, why does it ignore the other values? Further, in the backward pass the value of that index is copied to all inputs, irrespectively if it's the last or not. There is no checking if this is the last value where ```maxp``` is used.

I think either the forward pass needs to do: ```output_ptr[maxp] += input_data[i];``` so that all values are taken, and therefore broadcasting ```grad_output_ptr[maxp]``` to all grad_inputs that are using that index is correct, or the backward pass needs to filter which is the last ```i``` that uses ```maxp``` and only set the gradient for this last entry.

Unfortunately I have not found any reference which could explain how the forward pass should react on identical indices. Papers that use it like https://arxiv.org/pdf/1505.04366.pdf and https://arxiv.org/pdf/1311.2901.pdf seem only to use non-overlapping MaxPoolings (```kernel_size == stride```) in which this error does not occur. It only happens when ```kernel_size > stride```. And papers that have been released later use implementations from the AI frameworks.

TensorFlow does not natively support this operation, but it part of TF-Addons: ```https://github.com/tensorflow/addons/blob/v0.17.1/tensorflow_addons/layers/max_unpooling_2d.py#L46```

Here they use ```tf.scatter```, which sums up entries with duplicate index, see https://www.tensorflow.org/api_docs/python/tf/scatter_nd

**Comment by nikitaved** [comment]
@mergian , there a  couple of issues.

First, `scatter`, just as the forward implementation, is non-deterministic with repeated indices, as you cannot tell which thread writes first.
Using `+=`, which could be done with `scatter_add` is also problematic, because now we need synchronization to avoid race conditions. This synchronization will result in quite a perf-hit for CUDA Tensors.
Also, bacwards for `scatter` and `scatter_add` are broken with repeated indices for the very same reasons I am afraid :(

**Comment by Lezcano** [comment]
So, the underlying issue here is that `max` is not differentiable when the maximum value is not unique. I think what we should do here is to document this the `MaxPool*d` functions.

@mergian would you want to submit a fix for this?

**Comment by nikitaved** [comment]
Having a single value getting all the grad is a valid subgradient, actually. This behavior could get implemented if forward is aware of repeated indices in advance.

**Comment by mergian** [comment]
@Lezcano: yes I could prepare a fix for the documentation.

@nikitaved: I put some hours into how to fix this. I think we can distinguish 2 cases:

**Case 1:** Non-overlapping poolings, when ```(dilation * kernel_size - 1) <= stride```. In this case the current implementation is perfectly fine as there can't be any conflicts.

**Case 2:** In the other cases, we could use the MaxPoolBackward implementation. It's deterministic (on CPU and GPU) and accumulates the inputs. It would require a new MaxUnPoolBackward kernel, but that is something like:

```cpp
at::parallel_for(..., [&](...) {
	int batch = ...
	int channels = ...
	int y = ...
	int x = ...
	
	scalar_t sum = 0;
	int64_t oindex = batch * ...;
	
	for(int ky ...) {
		int64_t iy = y * stride_y + ky * dilation_y - padding_y;
		if(((uint64_t)iy) < input_y) {
			for(int kx ...) {
				int64_t ix = x * stride_x + kx * dilation_x;
				if(((uint64_t)ix) < input_x) {
					int64_t iindex = iy * input_x + ix;
					if(indices[oindex] == iindex)
						sum += input[iindex];
				}
			}
		}
	}
	
	output[oindex] = sum;
});
```

So the ```at::max_unpool2d``` callback could check the condition and then dispatch to the correct implementation.

I also thought about a way to filter duplicate entries in die indices, but it requires to store another copy of indices for the backward pass, increasing the memory consumption during training and I didn't come up with a fast and deterministic implementation. Either it was a 2-pass implementation, or required atomics on GPUs.

**Comment by nikitaved** [comment]
Sorry, @mergian , for the delay. Let me comprehend your message and I will get back to you.

**Comment by mergian** [comment]
Yes, I will take a look into how to fix this, when I find time.

**Comment by mergian** [comment]
FYI: I'm currently quite busy and soon on vacation, so this will take at least till end of September. In case someone else wants to submit something earlier.

**Comment by zou3519** [comment]
I ran into this today. It looks like max_unpool jvp and max_unpool double backwards are non-deterministic (the following produces different result every time you run it). This seems like a silent correctness issue (it should either be deterministic or we should document this as a non-deterministic operation), so tentatively bumping the priority.

```py
import torch
import torch.nn.functional as F
import torch.autograd.forward_ad as fwAD

x = torch.tensor([[[ 5.1576,  5.1576],[-2.7036, -2.1172]]], requires_grad=True)
indices = torch.tensor([[[1, 1], [1, 2]]])
kwargs = {'kernel_size': 3, 'stride': 2, 'padding': (1,), 'output_size': torch.Size([1, 2, 3])}

torch.manual_seed(0)
t = torch.randn_like(x)

def max_unpool1d(x):
  return F.max_unpool1d(x, indices, **kwargs)

def compute_jvp(x, t):
  with fwAD.dual_level():
    x_dual = fwAD.make_dual(x, t)
    y = max_unpool1d(x_dual)
    return fwAD.unpack_dual(y).tangent

compute_jvp(x, t)
```

**Comment by Lezcano** [comment]
I think marking and documenting it as non-deterministic should be enough here. CC @kurtamohler

**Comment by nikitaved** [comment]
Alas, no way about it as of now. The double backward relies on `gather`, and it is non-deterministic for repeated values.

**Comment by albanD** [comment]
>  The doc mentions no repeated values, as currently the grad propagates only to a single value non-deterministically

Does it? Or does it propagate to all inputs?


In any case, this should be marked as non-deterministic properly.

**Comment by nikitaved** [comment]
It should to all inputs as min/max is doing so, but it now it only propagates to a single value afair.

**Comment by albanD** [comment]
The min/max situation is a bit more subtle:
- If only one value in the input achieves the extremum, it gets the gradient
- If multiple inputs achieve the extremum AND the function returns the index of the extremum (for example t.max(dim=0) which picked one of them arbitrarily), then the input for that index gets the gradient
- If multiple inputs achieve the extremum AND the function does NOT return the indices (t.max() for example), the gradient is spread (divided by the number of elements achieving that extremum) across all inputs achieving the extremum

**Comment by jbschlosser** [comment]
Removing high-pri as the alert is now documented.

**Comment by drisspg** [comment]
Update will make sure that this function behaves correctly in deterministic modes

**Comment by drisspg** [comment]
https://github.com/pytorch/pytorch/blob/5c16788e5ff5ed1b3eba9c8fde5fc0910c495fa8/aten/src/ATen/native/MaxUnpooling.cpp#L25

There already appears the correct deterministic handiling for the CPU path.

**Comment by jbschlosser** [comment]
actionable to make sure this is documented

**Comment by shaoyuyoung** [comment]
It seem that this issue still exists on CPU, is there any subsequent plan?
